pkg/controller/node/ipam/cloud_cidr_allocator.go

```diff
@@ -42,84 +45,149 @@ import (
-
 	if len(cidrs) == 0 {
+		ca.removeNodeFromProcessing(node.Name)
 		util.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
-		glog.V(2).Infof("Node %v has no CIDRs", node.Name)
-		return fmt.Errorf("failed to allocate cidr (none exist)")
+		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
 
-	node, err = ca.client.Core().Nodes().Get(node.Name, metav1.GetOptions{})
-	if err != nil {
-		glog.Errorf("Could not get Node object from Kubernetes: %v", err)
-		return err
+	glog.V(10).Infof("Putting node %s with CIDR %s into the work queue", node.Name, cidrs[0])
+	ca.nodeCIDRUpdateChannel <- nodeAndCIDR{
+		nodeName: node.Name,
+		cidr:     cidrs[0],
 	}
+	return nil
+}
 
-	podCIDR := cidrs[0]
-
-	if node.Spec.PodCIDR != "" {
-		if node.Spec.PodCIDR == podCIDR {
-			glog.V(3).Infof("Node %v has PodCIDR %v", node.Name, podCIDR)
-			return nil
+// Assigns CIDR to Node and sends an update to the API server.
+func (ca *cloudCIDRAllocator) updateCIDRAllocation(data nodeAndCIDR) error {
+	var err error
+	var node *v1.Node
+	defer ca.removeNodeFromProcessing(data.nodeName)
+	podCIDR := data.cidr
+	for rep := 0; rep < cidrUpdateRetries; rep++ {
+		// TODO: change it to using PATCH instead of full Node updates.
+		node, err = ca.client.Core().Nodes().Get(data.nodeName, metav1.GetOptions{})
+		if err != nil {
+			glog.Errorf("Failed while getting node %v to retry updating Node.Spec.PodCIDR: %v", data.nodeName, err)
+			continue
+		}
+		if node.Spec.PodCIDR != "" {
+			if node.Spec.PodCIDR == podCIDR {
+				glog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDR)
+				return nil
+			}
+			glog.Errorf("PodCIDR cannot be reassigned, node %v spec has %v, but cloud provider has assigned %v",
+				node.Name, node.Spec.PodCIDR, podCIDR)
+			// We fall through and set the CIDR despite this error. This
+			// implements the same logic as implemented in the
+			// rangeAllocator.
+			//
+			// See https://github.com/kubernetes/kubernetes/pull/42147#discussion_r103357248
+		}
+		node.Spec.PodCIDR = podCIDR
+		if _, err = ca.client.Core().Nodes().Update(node); err != nil {
+			glog.Errorf("Failed to update node %v PodCIDR to %v (%d retries left): %v", node.Name, podCIDR, cidrUpdateRetries-rep-1, err)
+		} else {
+			break
 		}
-		glog.Errorf("PodCIDR cannot be reassigned, node %v spec has %v, but cloud provider has assigned %v",
-			node.Name, node.Spec.PodCIDR, podCIDR)
-		// We fall through and set the CIDR despite this error. This
-		// implements the same logic as implemented in the
-		// rangeAllocator.
-		//
-		// See https://github.com/kubernetes/kubernetes/pull/42147#discussion_r103357248
 	}
-
-	node.Spec.PodCIDR = cidrs[0]
-	if _, err := ca.client.Core().Nodes().Update(node); err == nil {
-		glog.V(2).Infof("Node %v PodCIDR set to %v", node.Name, podCIDR)
-	} else {
-		glog.Errorf("Could not update node %v PodCIDR to %v: %v",
-			node.Name, podCIDR, err)
+	if err != nil {
```

Comment:
Don't need to check `err` here, as we have going through the retry loop without success (if you change the above code to return)

Reply:
Didn't change the above code to return as we want to reach the code below where NodeCondition is set to `RouteCreated`. So have to explicitly check for the error.